keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -944,23 +948,77 @@ The linter, as previously described, will enforce rules to address valid zero-va
 
-```
+* They operate on the same underlying storage object as the primary resource (i.e., same `kind`, same object in etcd).
+* Updates via these subresources are typically scoped to specific fields within the object. Changes to field values not in scope are typically reset, or "wiped", before validation.
+* In some cases, they permit updates to fields that cannot be changed via the primary resource.
+
+**Examples:**
+
+* `pods/status`: Allows updates only to the `metadata` and `status` stanzas, prohibiting changes to the `spec`.
+* `resourceclaims/status`: Allows updates only to the `status` stanza, prohibiting changes to `metadata` or `spec`.
+* `pods/resize`: Allows updates to `spec.container[*].resources` fields, which are normally immutable after Pod creation via the primary `pods` resource.
+* `certificatesigningrequests/approval`: Allows adding, removing, or modifying Approve/Deny conditions within the `status`, which is not allowed via the
+   primary `certificatesigningrequests` resource nor by the `certificatesigningrequests/status` resource.
+
+**Validation Process:**
+
+1.  **Field Wiping (Pre-Validation):** Declarative Validation *does not* scope which fields a subresource operation can write. Instead, this responsibility lies with the resource's strategy implementation. The strategy modifies the incoming object *before* validation, "wiping" or resetting any fields the specific subresource operation is not allowed to change. Once wiped, the ratcheting mechanism (below) will skip validation of these fields, since wiping ensures that they are unchanged.
+2.  **Full Resource Validation:** The *entire*, modified resource object is validated against the primary resource's versioned type. That is, *same* set of declarative validation rules applied to the primary resource and to status-type subresources.
+3.  **Conditional Validation:** Validation rules can use a special `subresource` parameter (e.g., `subresource == '/status'`) to apply conditional logic. This allows rules to behave differently depending on whether the update comes via the primary resource or a specific subresource.
+4.  **Ratcheting:** Declarative Validation uses ratcheting, meaning validation does not fail for fields that have not changed from the existing stored object. Combined with field wiping, this scopes validation to only the subset of fields that the subresource operation is intended and permitted to modify.
+
+**Validation Examples:**
+
+* An update via `pods/status` first has its `spec` field changes wiped by the Pod strategy. Then, the entire Pod object is validated using the standard Pod validation rules. Ratcheting skips checks on unchanged `metadata` or `status` fields.
+* An update via `pods/resize` is validated using the standard Pod rules. However, a rule on `spec.container[*].resources` might look like `+k8s:if('subresource != "/resize"')=+k8s:immutable`, effectively enforcing immutability *unless* the update comes via the `resize` subresource.
```

Comment:
I think we need to formalize what the tag looks like.  I see access to which endpoint was used to change the resource as independent of generalized conditional logic.

Reply:
If we go with dedicated tags, the tags I've been considering are:

- `+k8s:subresource("/resize")=+k8s:xyz`
- `+k8s:exceptSubresource("/resize")=+k8s:immutable`

This would cover the vast majority of cases (all known cases, I believe).  We could still have the generalized conditional logic as an escape hatch.

Thoughts?

Reply:
I've thrown in `+k8s:ifSubresource` and `+k8s:ifNotSubresource` as placeholder for now.